torch/tensor.py

```diff
@@ -288,10 +288,10 @@ def trim(str):
 
     def retain_grad(self):
         r"""Enables .grad attribute for non-leaf Tensors."""
-        if self.grad_fn is None:  # no-op for leaves
-            return
         if not self.requires_grad:
             raise RuntimeError("can't retain_grad on Tensor that has requires_grad=False")
+        if self.is_leaf:  # no-op for leaves
```

Comment:
probably worth noting that this is BC-breaking, although this would be a weird thing to do.  But just add "label: bc-breaking" and show a simple example.

Reply:
The change to `gradcheck` is a good example of the kind of BC change.
Basically calling `.retain_grad()` on something that is going to be ignored later.